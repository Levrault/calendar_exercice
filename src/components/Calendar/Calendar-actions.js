import axios from 'axios';

export const CALENDAR_FETCH_BEGIN = 'CALENDAR_FETCH_BEGIN';
export const CALENDAR_FETCH_SUCCESS = 'CALENDAR_FETCH_SUCCESS';
export const CALENDAR_FETCH_FAILURE = 'CALENDAR_FETCH_FAILURE';

/**
 * Begin to fetch data
 * @returns {object}
 */
export const calendarFetchBegin = () => ({
  type: CALENDAR_FETCH_BEGIN
});

/**
 * Calendar data has been returned
 * @param {array} months
 * @returns {object}
 */
export const calendarFetchSuccess = months => ({
  type: CALENDAR_FETCH_SUCCESS,
  payload: {
    months
  }
});

/**
 * Receive an error
 * @param {object} error
 * @returns {object}
 */
export const calendarFetchError = error => ({
  type: CALENDAR_FETCH_FAILURE,
  payload: { error }
});

export const fetch = (params) => async (dispatch) => {
  dispatch(calendarFetchBegin());
  try {
    const response = await axios.get(`${BASE_URL}calendar`, {
      params
    });
    dispatch(calendarFetchSuccess(response.data[0].agenda));
  } catch (error) {
    dispatch(calendarFetchError(error));
  }
};
