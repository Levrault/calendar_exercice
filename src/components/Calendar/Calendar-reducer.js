import {
  CALENDAR_FETCH_BEGIN,
  CALENDAR_FETCH_SUCCESS,
  CALENDAR_FETCH_FAILURE
} from './Calendar-actions';

const initialState = {
  months: [],
  loading: false,
  error: null
};

export const calendarReducer = (state = initialState, action) => {
  switch (action.type) {
  case CALENDAR_FETCH_BEGIN:
    return {
      ...state,
      loading: true,
      error: null
    };

  case CALENDAR_FETCH_SUCCESS:
    return {
      ...state,
      loading: false,
      ...action.payload
    };

  case CALENDAR_FETCH_FAILURE:
    return {
      ...state,
      loading: false,
      error: action.payload.error,
      calendar: []
    };

  default:
    return state;
  }
};
