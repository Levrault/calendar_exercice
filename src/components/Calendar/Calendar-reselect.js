import { createSelector } from 'reselect';
import get from 'lodash/get';

const calendarSelector = state => get(state, 'calendar');

export const monthsSelector = createSelector(
  calendarSelector,
  calendar => get(calendar, 'months', [])
);

export const loadingSelector = createSelector(
  calendarSelector,
  calendar => get(calendar, 'loading', false)
);

export const errorSelector = createSelector(
  calendarSelector,
  calendar => get(calendar, 'error', null)
);

