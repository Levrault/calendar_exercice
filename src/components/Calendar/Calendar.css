.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.title {
  color: $palette-title;
  font-weight: 300;
  margin: $space-l auto;
}

.content {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  position: relative;
}
