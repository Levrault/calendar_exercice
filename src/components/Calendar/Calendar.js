import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MonthConnected from '../Month/MonthConnected';
import Loader from '../Loader/Loader';
import Error from '../Errors/Error';
import styles from './Calendar.css';

/**
 * Calendar layout
 * @class Calendar
 * @param {number} year
 * @param {array} months
 * @param {bool} [isLoading=true]
 * @param {object|null} [error]
 */
class Calendar extends Component {
  /**
   * Update only when news months are receive
   * @param {object} nextProps
   * @returns {bool}
   */
  shouldComponentUpdate (nextProps) {
    return nextProps.months.length > 0 && nextProps.months !== this.props.months;
  }

  /**
  * Render if there's no errors and
  * when months are not empty
  * @returns {node}
  */
  render () {
    const { months, isLoading, error, year } = this.props;

    if (error !== null) {
      return <Error {...error} />;
    }

    if (!months.length) {
      return null;
    }

    return (
      <div className={styles.container}>
        <h1 className={styles.title}>Calendar {year}</h1>
        <div className={styles.content}>
          {
            months.map(month => (
              <MonthConnected key={month.name} {...month} year={year} />
            ))
          }
          {isLoading &&
            <Loader />
          }
        </div>
      </div>
    );
  }
}

Calendar.propTypes = {
  year: PropTypes.number.isRequired,
  months: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    numberOfDays: PropTypes.number.isRequired,
    chronology: PropTypes.number.isRequired,
    firstDayIndex: PropTypes.number.isRequired
  })),
  isLoading: PropTypes.bool,
  error: PropTypes.object
};

Calendar.defaultProps = {
  year: 2018,
  months: [],
  isLoading: true
};

export default Calendar;
