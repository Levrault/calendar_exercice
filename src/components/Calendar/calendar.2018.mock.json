{
 "agenda": [{
    "name": "January",
    "numberOfDays": 31,
    "year": 2018,
    "monthNumber": 1,
    "firstDayOfMonth": 2,
    "events": ["2", "15"]
  },
  {
    "name": "February",
    "numberOfDays": 28,
    "year": 2018,
    "monthNumber": 2,
    "firstDayOfMonth": 6,
    "events": []
  },
  {
    "name": "March",
    "numberOfDays": 31,
    "year": 2018,
    "monthNumber": 3,
    "firstDayOfMonth": 1,
    "events": []
  },
  {
    "name": "April",
    "numberOfDays": 30,
    "year": 2018,
    "monthNumber": 4,
    "firstDayOfMonth": 2,
    "events": []
  },
  {
    "name": "May",
    "numberOfDays": 31,
    "year": 2018,
    "monthNumber": 5,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "June",
    "numberOfDays": 30,
    "year": 2018,
    "monthNumber": 6,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "July",
    "numberOfDays": 31,
    "year": 2018,
    "monthNumber": 7,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "August",
    "numberOfDays": 31,
    "year": 2018,
    "monthNumber": 8,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "September",
    "numberOfDays": 30,
    "year": 2018,
    "monthNumber": 9,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "October",
    "numberOfDays": 31,
    "year": 2018,
    "monthNumber": 10,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "November",
    "numberOfDays": 30,
    "year": 2018,
    "monthNumber": 11,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "December",
    "numberOfDays": 31,
    "year": 2018,
    "monthNumber": 12,
    "firstDayOfMonth": 5,
    "events": []
  }
]
}