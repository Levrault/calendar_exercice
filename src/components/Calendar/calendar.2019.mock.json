{
 "agenda": [{
    "name": "January",
    "numberOfDays": 31,
    "year": 2019,
    "monthNumber": 1,
    "firstDayOfMonth": 1,
    "events": []
  },
  {
    "name": "February",
    "numberOfDays": 28,
    "year": 2019,
    "monthNumber": 2,
    "firstDayOfMonth": 2,
    "events": []
  },
  {
    "name": "March",
    "numberOfDays": 31,
    "year": 2019,
    "monthNumber": 3,
    "firstDayOfMonth": 3,
    "events": []
  },
  {
    "name": "April",
    "numberOfDays": 30,
    "year": 2019,
    "monthNumber": 4,
    "firstDayOfMonth": 2,
    "events": []
  },
  {
    "name": "May",
    "numberOfDays": 31,
    "year": 2019,
    "monthNumber": 5,
    "firstDayOfMonth": 4,
    "events": []
  },
  {
    "name": "June",
    "numberOfDays": 30,
    "year": 2019,
    "monthNumber": 6,
    "firstDayOfMonth": 2,
    "events": []
  },
  {
    "name": "July",
    "numberOfDays": 31,
    "year": 2019,
    "monthNumber": 7,
    "firstDayOfMonth": 3,
    "events": []
  },
  {
    "name": "August",
    "numberOfDays": 31,
    "year": 2019,
    "monthNumber": 8,
    "firstDayOfMonth": 4,
    "events": []
  },
  {
    "name": "September",
    "numberOfDays": 30,
    "year": 2019,
    "monthNumber": 9,
    "firstDayOfMonth": 5,
    "events": []
  },
  {
    "name": "October",
    "numberOfDays": 31,
    "year": 2019,
    "monthNumber": 10,
    "firstDayOfMonth": 2,
    "events": []
  },
  {
    "name": "November",
    "numberOfDays": 30,
    "year": 2019,
    "monthNumber": 11,
    "firstDayOfMonth": 1,
    "events": []
  },
  {
    "name": "December",
    "numberOfDays": 31,
    "year": 2019,
    "monthNumber": 12,
    "firstDayOfMonth": 5,
    "events": []
  }
]
}