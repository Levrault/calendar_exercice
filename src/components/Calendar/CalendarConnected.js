import { connect } from 'react-redux';
import Calendar from './Calendar';
import {
  loadingSelector,
  errorSelector
} from '../YearForm/YearForm-reselect';

/**
 * @param {object} state
 * @returns {object}
 */
const mapStateToProps = (state) => ({
  isLoading: loadingSelector(state),
  error: errorSelector(state)
});

export default connect(mapStateToProps)(Calendar);
