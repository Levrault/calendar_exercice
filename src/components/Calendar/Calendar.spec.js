import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import CalendarConnected from './CalendarConnected';
import Calendar from './Calendar';
import getFakeStore from '../../tools/test/fakeStore';
import getFakeData from '../../tools/test/fakeData';
import { shallow } from '../../tools/test/enzyme';

const store = getFakeStore();
const mock = getFakeData().year.current;
const nextMock = getFakeData().year.data[1];

describe('Calendar component', () => {
  it('should render without problem', () => {
    const component = renderer.create(
      <Provider store={store}>
        <CalendarConnected />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return an error when there is an error', () => {
    const component = renderer.create(
      <Calendar {...mock} error={{ response: { status: 404 } }} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return null when there is no month', () => {
    const component = shallow(
      <Calendar {...mock} months={[]} error={null} />
    );

    expect(component.isEmptyRender()).toBeTruthy();
  });

  it('should render only when there is new months has props', () => {
    const component = shallow(
      <Calendar {...mock} error={null} />
    );

    expect(component.instance().shouldComponentUpdate(nextMock)).toBeTruthy();
    expect(component.instance().shouldComponentUpdate({ months: [] })).toBeFalsy();
  });
});
