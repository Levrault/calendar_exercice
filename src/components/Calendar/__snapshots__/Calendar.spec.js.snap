// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component should render without problem 1`] = `null`;

exports[`Calendar component should return an error when there is an error 1`] = `
<div>
  <div>
    404 not found
  </div>
</div>
`;
